librz/bin/format/pe/pe_relocs.c

```diff
@@ -0,0 +1,72 @@
+// SPDX-FileCopyrightText: 2024 Roee Toledano <roeetoledano10@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include "pe.h"
+#include <rz_util/ht_uu.h>
+
+static unsigned long read_reloc_ent_from_block(RZ_NONNULL RzVector /*<RzBinPeRelocEnt>*/ *relocs, RzBuffer *b, int big_endian, ut32 block_size) {
+	unsigned int bytes_read = sizeof(RzBinPeRelocBlock);
+
+	do {
+		bytes_read += 2;
+		RzBinPeRelocEnt reloc;
+		ut8 ent_buf[sizeof(RzBinPeRelocEnt)];
+		rz_buf_read(b, ent_buf, sizeof(ent_buf));
+		reloc = rz_read_ble16(ent_buf, big_endian);
+
+		if (reloc == 0) { // we reached last 2 bytes of the block. these are zeroed padding bytes
+			break;
+		}
+		// encoding is 12 bits for type and 4 bits for offset. See: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#the-reloc-section-image-only
+		rz_vector_push(relocs, &reloc);
+
+	} while (bytes_read < block_size);
+
+	return bytes_read;
+}
+
+static bool get_relocs_from_data_dir(RZ_NONNULL RzBinPEObj *bin, RZ_NONNULL RzVector /*<RzBinPeRelocEnt>*/ *relocs) {
+	RzBuffer *b = bin->b;
+	st64 o_addr = rz_buf_tell(b);
+	{ // get offset in file of first reloc block
+		PE_DWord offset = PE_(bin_pe_rva_to_paddr)(bin, bin->nt_headers->optional_header.DataDirectory[PE_IMAGE_DIRECTORY_ENTRY_BASERELOC].VirtualAddress);
+		if (rz_buf_seek(b, offset, RZ_BUF_SET) < 0) {
+			return false;
+		}
+	}
+	unsigned long total_bytes_to_read = bin->nt_headers->optional_header.DataDirectory[PE_IMAGE_DIRECTORY_ENTRY_BASERELOC].Size;
+	unsigned long total_bytes_read = 0;
+
+	do {
+		ut8 block_buf[sizeof(RzBinPeRelocBlock)];
+		rz_buf_read(b, block_buf, sizeof(block_buf));
+		RzBinPeRelocBlock block;
+		PE_READ_STRUCT_FIELD_L(block_buf, block, RzBinPeRelocBlock, page_rva, 32);
+		PE_READ_STRUCT_FIELD_L(block_buf, block, RzBinPeRelocBlock, block_size, 32);
```

Comment:
you can do the same here using `rz_buf_read_bleXX_offset` and checking the return value.
you can do this in a static function like [this](https://github.com/rizinorg/rizin/blob/dev/librz/bin/p/bin_qnx.c#L7) and fail when returns false.